File: build_python_exercise/pyspark_examples/wordcount.py
#!/usr/bin/env python3
"""
PySpark WordCount Example
Count word frequencies in text files
"""

from pyspark.sql import SparkSession
from pyspark.sql.functions import explode, split, lower, regexp_replace, col
import sys

def main():
    # Create SparkSession
    spark = SparkSession.builder \
        .appName("WordCount") \
        .master("local[*]") \
        .getOrCreate()
    
    # Read input file
    input_file = sys.argv[1] if len(sys.argv) > 1 else "../data/sample_text.txt"
    
    # Read data
    df = spark.read.text(input_file)
    
    # Process text: lowercase, remove special chars, split words
    words = df.select(
        explode(
            split(
                regexp_replace(lower(col("value")), "[^a-zA-Z0-9\\s]", ""), 
                "\\s+"
            )
        ).alias("word")
    ).filter(col("word") != "")
    
    # Count word frequencies
    word_counts = words.groupBy("word").count().orderBy(col("count").desc())
    
    # Display results
    print("Top 20 most frequent words:")
    word_counts.show(20)
    
    # Save results
    output_path = "output/wordcount_result"
    word_counts.coalesce(1).write.mode("overwrite").csv(output_path, header=True)
    print(f"Results saved to: {output_path}")
    
    spark.stop()

if __name__ == "__main__":
    main()

File: build_scala_exercise/spark_sbt/src/main/scala/com/example/WordCount.scala
package com.example

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._

/**
 * Spark WordCount application demonstrating basic text processing
 */
object WordCount {
  
  def main(args: Array[String]): Unit = {
    // Create Spark session
    val spark = SparkSession.builder()
      .appName("WordCount")
      .config("spark.sql.adaptive.enabled", "true")
      .config("spark.sql.adaptive.coalescePartitions.enabled", "true")
      .getOrCreate()
    
    import spark.implicits._
    
    try {
      // Input and output paths
      val inputPath = if (args.length > 0) args(0) else "input.txt"
      val outputPath = if (args.length > 1) args(1) else "output"
      
      println(s"Processing file: $inputPath")
      println(s"Output directory: $outputPath")
      
      // Read text file
      val textDF = spark.read.text(inputPath)
      
      // Perform word count
      val wordCountDF = textDF
        .select(explode(split(col("value"), "\\s+")).as("word"))
        .filter(col("word") =!= "")
        .groupBy("word")
        .count()
        .orderBy(desc("count"))
      
      // Show results
      println("Top 20 words:")
      wordCountDF.show(20)
      
      // Save results
      wordCountDF
        .coalesce(1)
        .write
        .mode("overwrite")
        .option("header", "true")
        .csv(outputPath)
      
      println(s"Results saved to: $outputPath")
      
      // Print statistics
      val totalWords = wordCountDF.agg(sum("count")).collect()(0).getLong(0)
      val uniqueWords = wordCountDF.count()
      
      println(s"Total words: $totalWords")
      println(s"Unique words: $uniqueWords")
      
    } catch {
      case e: Exception =>
        println(s"Error processing file: ${e.getMessage}")
        e.printStackTrace()
        System.exit(1)
    } finally {
      spark.stop()
    }
  }
}